diff --git a/testlib.h b/testlib.h
index c45e320..17e39f2 100644
--- a/testlib.h
+++ b/testlib.h
@@ -60,6 +60,8 @@
  *   Reads test from inf (mapped to args[1]), writes result to tout (mapped to argv[2],
  *   can be judged by checker later), reads program output from ouf (mapped to stdin),
  *   writes output to program via stdout (use cout, printf, etc).
+ *
+ *   NOTE: This file is modified to run in domjudge.
  */
 
 const char *latestFeatures[] = {
@@ -234,7 +236,7 @@ const char *latestFeatures[] = {
 #   ifdef CONTESTER
 #       define OK_EXIT_CODE 0xAC
 #   else
-#       define OK_EXIT_CODE 0
+#       define OK_EXIT_CODE 42
 #   endif
 #endif
 
@@ -244,7 +246,7 @@ const char *latestFeatures[] = {
 #   elif defined(CONTESTER)
 #       define WA_EXIT_CODE 0xAB
 #   else
-#       define WA_EXIT_CODE 1
+#       define WA_EXIT_CODE 43
 #   endif
 #endif
 
@@ -254,7 +256,7 @@ const char *latestFeatures[] = {
 #   elif defined(CONTESTER)
 #       define PE_EXIT_CODE 0xAA
 #   else
-#       define PE_EXIT_CODE 2
+#       define PE_EXIT_CODE 43
 #   endif
 #endif
 
@@ -272,7 +274,7 @@ const char *latestFeatures[] = {
 #   ifdef EJUDGE
 #       define DIRT_EXIT_CODE 6
 #   else
-#       define DIRT_EXIT_CODE 4
+#       define DIRT_EXIT_CODE 43
 #   endif
 #endif
 
@@ -281,7 +283,7 @@ const char *latestFeatures[] = {
 #endif
 
 #ifndef UNEXPECTED_EOF_EXIT_CODE
-#   define UNEXPECTED_EOF_EXIT_CODE 8
+#   define UNEXPECTED_EOF_EXIT_CODE 43
 #endif
 
 #ifndef PC_BASE_EXIT_CODE
@@ -3081,7 +3083,7 @@ NORETURN void InStream::quit(TResult result, const char *msg) {
     }
 
     quitscr(LightGray, __testlib_toPrintableMessage(message).c_str());
-    std::fprintf(stderr, "\n");
+    // std::fprintf(stderr, "\n");
 
     inf.close();
     ouf.close();
@@ -3090,10 +3092,10 @@ NORETURN void InStream::quit(TResult result, const char *msg) {
         tout.close();
 
     textColor(LightGray);
-
+/*
     if (resultName != "")
         std::fprintf(stderr, "See file to check exit message\n");
-
+*/
     halt(resultExitCode(result));
 }
 
@@ -4496,6 +4498,11 @@ void setAppesModeEncoding(std::string appesModeEncoding) {
     ::appesModeEncoding = appesModeEncoding;
 }
 
+std::string make_new_file_in_a_dir(std::string dir, std::string file = "judgemessage.txt") { // assume in linux
+    if (dir.back() != '/') dir.push_back('/');
+    return dir + file;
+}
+
 void registerInteraction(int argc, char *argv[]) {
     __testlib_ensuresPreconditions();
     __testlib_set_testset_and_group(argc, argv);
@@ -4506,24 +4513,23 @@ void registerInteraction(int argc, char *argv[]) {
 
     if (argc > 1 && !strcmp("--help", argv[1]))
         __testlib_help();
-
+/*
     if (argc < 3 || argc > 6) {
         quit(_fail, std::string("Program must be run with the following arguments: ") +
                     std::string("<input-file> <output-file> [<answer-file> [<report-file> [<-appes>]]]") +
                     "\nUse \"--help\" to get help information");
     }
-
-    if (argc <= 4) {
-        resultName = "";
+*/
+    if (argc == 4) {
+        resultName = make_new_file_in_a_dir(argv[3]);
+        tout.open(make_new_file_in_a_dir(argv[3], "nextpass.in"), std::ios_base::out);
+        if (tout.fail() || !tout.is_open())
+            quit(_fail, std::string("Can not write to the test-output-file '") + argv[2] + std::string("'"));
         appesMode = false;
     }
 
 #ifndef EJUDGE
-    if (argc == 5) {
-        resultName = argv[4];
-        appesMode = false;
-    }
-
+/*
     if (argc == 6) {
         if (strcmp("-APPES", argv[5]) && strcmp("-appes", argv[5])) {
             quit(_fail, std::string("Program must be run with the following arguments: ") +
@@ -4533,18 +4539,19 @@ void registerInteraction(int argc, char *argv[]) {
             appesMode = true;
         }
     }
+*/
 #endif
 
     inf.init(argv[1], _input);
-
+/*
     tout.open(argv[2], std::ios_base::out);
     if (tout.fail() || !tout.is_open())
         quit(_fail, std::string("Can not write to the test-output-file '") + argv[2] + std::string("'"));
 
+*/
     ouf.init(stdin, _output);
-
-    if (argc >= 4)
-        ans.init(argv[3], _answer);
+    if (argc >= 3)
+        ans.init(argv[2], _answer);
     else
         ans.name = "unopened answer stream";
 }
@@ -4699,19 +4706,20 @@ void registerTestlibCmd(int argc, char *argv[]) {
     argc = int(args.size());
     if (argc > 1 && "--help" == args[1])
         __testlib_help();
-
+/*
     if (argc < 4 || argc > 6) {
         quit(_fail, std::string("Program must be run with the following arguments: ") +
                     std::string("[--testset testset] [--group group] <input-file> <output-file> <answer-file> [<report-file> [<-appes>]]") +
                     "\nUse \"--help\" to get help information");
     }
-
+*/
     if (argc == 4) {
-        resultName = "";
+        resultName = make_new_file_in_a_dir(args[3]);
         appesMode = false;
     }
 
 #ifndef EJUDGE
+/*
     if (argc == 5) {
         resultName = args[4];
         appesMode = false;
@@ -4726,12 +4734,13 @@ void registerTestlibCmd(int argc, char *argv[]) {
             appesMode = true;
         }
     }
+*/
 #endif
 
     inf.init(args[1], _input);
-    ouf.init(args[2], _output);
+    ouf.init(stdin, _output);
     ouf.skipBom();
-    ans.init(args[3], _answer);
+    ans.init(args[2], _answer);
 }
 
 void registerTestlib(int argc, ...) {
